toolkit/tools/pkg/imagecustomizerlib/cosicommon.go

```diff
@@ -172,21 +211,119 @@ func sha384sum(path string) (string, error) {
+		return fmt.Errorf("failed to calculate sha384 of %s:\n%w", source, err)
+	}
+	imageFile.Sha384 = sha384
+
+	return nil
+}
+
+func populateMetadata(data ImageBuildData) error {
+	if err := populateImageFile(data.Source, &data.Metadata.Image); err != nil {
+		return fmt.Errorf("failed to populate metadata: %w", err)
+	}
+
+	if err := populateVerityMetadata(data.VeritySource, data.Metadata.Verity); err != nil {
+		return fmt.Errorf("failed to populate verity metadata: %w", err)
+	}
+
+	return nil
+}
+
+func populateVerityMetadata(source *string, verityList []*Verity) error {
+	if source == nil && len(verityList) == 0 {
+		return nil
+	}
+
+	if source == nil || len(verityList) == 0 {
+		return fmt.Errorf("source and verity list must be both defined or both undefined")
+	}
+
+	for _, verity := range verityList {
+		if verity == nil {
+			continue
+		}
+
+		if err := populateImageFile(*source, &verity.Image); err != nil {
+			return fmt.Errorf("failed to populate verity image metadata: %w", err)
+		}
 	}
-	data.Metadata.Image.Sha384 = sha384
+
 	return nil
 }
+
+func buildCosiData(partitions []outputPartitionMetadata, verityMetadata []verityDeviceMetadata) ([]cosiPartitionMetadata, error) {
+	var cosiMetadata []cosiPartitionMetadata
+
+	// Pre-compute a map for quick lookup of partition metadata by UUID
+	type PartitionMetadata struct {
+		Filename         string
+		UncompressedSize uint64
+	}
+
+	partUuidToMetadata := make(map[string]PartitionMetadata)
+	for _, partition := range partitions {
+		partUuidToMetadata[partition.PartUuid] = PartitionMetadata{
+			Filename:         partition.PartitionFilename,
+			UncompressedSize: partition.UncompressedSize,
+		}
+	}
+	for _, partition := range partitions {
+		// Skip roothash partitions
+		if partition.FileSystemType == "DM_verity_hash" {
+			continue
+		}
+
+		cosiPartition := cosiPartitionMetadata{
+			PartitionNum:      partition.PartitionNum,
+			PartitionFilename: partition.PartitionFilename,
+			PartLabel:         partition.PartLabel,
+			FileSystemType:    partition.FileSystemType,
+			PartitionTypeUuid: partition.PartitionTypeUuid,
+			Uuid:              partition.Uuid,
+			PartUuid:          partition.PartUuid,
+			Mountpoint:        partition.Mountpoint,
+			UncompressedSize:  partition.UncompressedSize,
+			Verity:            nil,
+		}
+
+		for _, verity := range verityMetadata {
+			if partition.PartUuid == verity.dataPartUuid {
+				metadata, exists := partUuidToMetadata[verity.hashPartUuid]
+				if !exists || metadata.Filename == "" {
+					return nil, fmt.Errorf("missing metadata for hash partition UUID: %s", verity.hashPartUuid)
+				}
+
+				// Create Verity entry
+				verityEntry := &Verity{
+					Hash: verity.hash,
+					Image: ImageFile{
+						Path:             metadata.Filename,
+						UncompressedSize: metadata.UncompressedSize,
+					},
+				}
+
+				if cosiPartition.Verity == nil {
+					cosiPartition.Verity = []*Verity{}
```

Comment:
In Go, `nil` is a valid value for an array. It represents an empty array. So, you don't need to do this.